### DISCUSSION TITLE
Woodgate Algorithm

### DISCUSSION THREAD

**Discussion by banrovegrie** [theochem/procrustes#171] [General]
Hey, this is wrt #170. I had a few doubts. @FanwangM  @Ali-Tehrani

**Comment by banrovegrie** [comment]
## Doubt 1

We know that $f$ is defined as follows in the paper. Now, if $F, G \in \mathbb{R}^{n\times m}$ then $F'F \in \mathbb{R}^{m\times m}$. As a result if $n \neq m$ then how will the dimensions of $F'F,\ (E'E)^2GG'$ match? Of course, $E \in \mathbb{R}^{n\times n}$.

<img src="https://user-images.githubusercontent.com/52317522/177386303-954faafb-d540-4bab-bf67-c986ad2e5ecf.png" width="420">

**Reply by FanwangM** [reply]
It seems that $n=m$ in order to make this equation true. 

What do you think? @Ali-Tehrani

**Reply by Ali-Tehrani** [reply]
The formula could have been written better.  The trace is invariant under the cyclic [permutations](https://math.stackexchange.com/questions/252272/is-trace-invariant-under-cyclic-permutation-with-rectangular-matrices) (and also transpose),  so you can just change the order so that the matrix shapes all match.

Alternately,  use the fact that the trace is linear and do `(tr(F`F) + Tr((E"E)^2 G G") - Tr(E"EQ))/2` separately. The latter is more efficient.

**Reply by banrovegrie** [reply]
I started working with it by changing the order but taking the traces separately is so much better, thankss.

**Comment by banrovegrie** [comment]
## Doubt 2

To solve the following minimization problem, I used `scipy.optimize.minimize`. However, I am confused as to how do I consider the fact that $\omega > 0$.

<img src="https://user-images.githubusercontent.com/52317522/177388715-2821b124-e279-435f-813e-85aaa56b7233.png" width="420">

Currently, I have put the bounds as follows and I think `(0, None)` is considered to be $\geq 0$.

```py

def find_minima(E, DE):
    def func(w):
        return f(E - w * DE)

    # Optimize the function using scipy.optimize.minimize
    # with respect to w > 0.
    w_min = minimize(func, 1, bounds=((0, None),))
    return w_min.x[0]
```

**Reply by FanwangM** [reply]
Maybe adding constraints with the following example, https://stackoverflow.com/questions/20075714/scipy-minimize-with-constraints?

**Reply by Ali-Tehrani** [reply]
I'm fine with replacing zero with a small number 1e-10.

**Reply by banrovegrie** [reply]
Yep I kept it 1e-9

**Comment by banrovegrie** [comment]
## Doubt 3

The error that we get once the algorithm convergences seems to be dependent on the initial choice of $E_0$ to some extent. Does this mean something is going wrong in the computation or else how should we choose a suitable $E_0$?

**Reply by FanwangM** [reply]
I am not quite sure about this. But maybe you are right, the algorithm is sensitive to the initialization. Do you have an example?

**Reply by banrovegrie** [reply]
I think I can fix this with some linear scaling.

**Reply by Ali-Tehrani** [reply]
If by linear scaling, you mean the alpha in the paper (or the N() function) then I agree with you.  Unless I missed it, the code also doesn't seem to scale by alpha at the end of the iteration either.

**Comment by FanwangM** [comment]
In order to render the equations properly, one will need https://github.com/nschloe/xhub.

**Reply by FanwangM** [reply]
I have a conference coming tomorrow and I will try to answer your questions tomorrow. Sorry for the delay.

**Reply by banrovegrie** [reply]
Ayy no no that is fine. You can take a look whenever u are free.
